text/0000-named-arguments.md

```diff
@@ -0,0 +1,623 @@
+/// ```
+fn split(string: &str, .at: char, .limit: usize, .case_sensitive: bool) {}
+
+// a dependent crate using the 2018 edition has to write
+split("hello world", ' ', 2, true);
+```
+
+If this function was written today, it might be named `split_at_with_limit` instead, and an enum would be used instead of a `bool`, so it would be readable without named arguments. However, since it was designed with named arguments in mind, the function name is unspecific, making it very difficult to understand without argument names.
+
+Therefore I believe that named arguments should be supported in all editions.
+
+
+## Alternative: Make named arguments in function calls mandatory
+
+See the _drawbacks_ section above.
+
+## Alternative: Do nothing
+
+There are many ways how code can be made more readable in the absence of named arguments. However, I consider them "hacks" and "workarounds", whereas named arguments is in many cases the most elegant, expressive and readable solution.
+
+### Parameter hints in IDEs
+
+IDEs can provide hints for function arguments. Both IntelliJ-Rust (the plugin that provides Rust language support in JetBrains IDEs) and rust-analyzer (a Rust language server that is going to replace RLS) can do this. Many Rust programmers use these IDE-powered parameter hints, which proves that there is demand for named arguments, at least when reading code.
+
+Unfortunately, parameter hints are not available when reading code in a blog post, in an online forum, on GitHub or any other website. Furthermore, some editors aren't able to show parameter hints.
+
+Another problem is that IDEs are not perfect. For example, when reading code in a `#[cfg(target_os = "windows")]` module on a Linux machine, neither IntelliJ-Rust nor rust-analyzer can provide parameter hints.
+
+Therefore, I believe that Rust code shouldn't rely on an IDE to be readable and expressive.
+
+### Comments
+
+For example:
+
+```rust
+free_fall(/*z0*/ 100.0, /*v0*/ 0.0, /*g*/ 9.81);
+
+free_fall(100.0, 0.0, 9.81); // z0, v0, g
+
+free_fall(
+    100.0,  // z0
+    0.0,    // v0
+    9.81,   // g
+);
+```
+
+Since Rust code like this appears very rarely in the wild, it appears that most Rust programmers either aren't aware of this possibility, or don't like writing code like this.
+
+### Enums for `bool`
+
+Instead of a `bool` argument, a custom enum with descriptive variant names can be created, for example:
+
+```rust
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum CaseSensitivity {
+    CaseSensitive,
+    CaseInsensitive,
+}
+```
+
+This improves type safety, but it also has disadvantages:
+
+ * It adds boilerplate for the library author
+ * The custom enum must be `use`d at every call site, which also adds boilerplate for the library user
+ * It doesn't implement every trait `bool` implements
+ * Using it is more verbose. For example, `if is_case_sensitive {}` might become `if case_sensitivity == CaseSensitivity::CaseSensitive {}`
```

Comment:
I think this section overly argues against two-variant enums without clarifying their advantages, and also moving into shaky territory.

* Boilerplate is generally _not_ considered a huge problem for large codebases, which the RFC uses as a fundamental base. On a large codebase, such definitions will most likely be the smallest part.
* The implementation presented is written very quickly.
* While it is true that this enum does not implement bool traits, most of the bool traits are binary operations, which will most likely not be used in this case. Other traits are easily derived.
* Verbosity is described as a disadvantage, which is debatable.

I'm not sure if this adds to the RFC enough.

Reply:
@skade - As someone who dropped Java for Scala precisely because of boilerplate in large codebases, I strongly disagree with your "boilerplate is generally not considered a huge problem".  It's a terrible, dreadful problem, leading to code development that slows to a crawl because you spend all your time finding things and reading things and typing things, and very little time in comparison actually thinking.  (Because of the huge burden of finding/reading/typing compared to how hard the thinking is once you actually wade through it all.)  Even people who claim to _like_ it still move like molasses and make lots of mistakes, in my observation.

Reply:
That's a good point. But I think this highly hinges on the nature of boilerplate. I would not use a 2-State enum as an example. It may be an example of verbosity, but not necessarily of boilerplate.

Rust shares a number of boilerplate problems with other languages (e.g. the extremely rote nature of delegation), but I'm not sure if the example given above fits the description.

I retreat the point that boilerplate is in general not problematic. I'm just not sure if the example carries enough weight to fall under that category.